### DISCUSSION TITLE
How to create a local nuget package and use it in my another project?

### DISCUSSION THREAD

**Discussion by BowenYang666** [dotnet/roslyn#75167] [General]
I have a source generator project called metricsFactoryTest which will generate some factory code:
![image](https://github.com/user-attachments/assets/44c5c7eb-9aa2-4060-8ad1-e1903cfdeb8f)

Now I want to reference this project in my another project called 'MainProject'.

One easy way to do this is to add a ProjectReference in the MainProject and this works for me
`<ProjectReference Include="..\MetricsFactoryTest\MetricsFactoryTest.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" /> -->`

But I want to publish the metricsFactoryTest to nuget package in the future and want to test locally if this package can work.

So I create a nuget package based on this guide: https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md#package-a-generator-as-a-nuget-package

here is my metricsFactoryTest.csproj file
![image](https://github.com/user-attachments/assets/4cb99be7-a210-4880-b614-5d0402c12465)

and then I create a package using 'dotnet pack' cmd to generate
MetricsFactoryTest.1.0.0.nupkg

After that I install this local package through visual studio in my MainProject:
![image](https://github.com/user-attachments/assets/4d38257a-5e6b-4081-9c0c-a33b03885fa7)

Then a package reference is generated in my MainProject.csproj
`<PackageReference Include="MetricsFactoryTest" Version="1.0.0" />`

but thenMetricsFactoryTest will not work, I cannot see this analyzers in the visual studio and I cannot use the code generated.
![image](https://github.com/user-attachments/assets/c6b2d6d0-1d45-490f-949d-46d2daf4276e)

if I switch back to project reference `<ProjectReference Include="..\MetricsFactoryTest\MetricsFactoryTest.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" />`

the analyzer will show up again
![image](https://github.com/user-attachments/assets/f60e461c-48da-43ee-90d7-6ad3f63bc65e)

So is there anything wrong with the my packaging steps? Not sure if I miss anything

**Comment by 333fred** [comment]
I would start by looking in the nuget package and making sure that everything is where you expect it to be.

On a side note, are you using Unity? Unless that's true, please don't use `ISourceGenerator`; you should be using `IIncrementalGenerator` instead, see https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md and https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.cookbook.md.

**Reply by BowenYang666** [reply]
I'm not using unity, here is the package version for generator class
![image](https://github.com/user-attachments/assets/d6fb16c3-13d5-4004-8472-f5c37b2d2a63)

**Reply by BowenYang666** [reply]
I wrote the generator class based on this guide, so it should be ok
https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview

**Reply by 333fred** [reply]
It is not ok. Please migrate to IIncrementalGenerator.

**Reply by 333fred** [reply]
@BillWagner, we should really mark that page deprecated.

**Reply by WeihanLi** [reply]
@333fred besides the document, seemed it's not obsoleted now from code base, is it applicable to mark the `ISourceGenerator` as obsolete and provide a message to move to the `IIncrementalGenerator`

https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/SourceGeneration/ISourceGenerator.cs#L22

```diff
+ [Obsolete("Please move to IIncrementalGenerator, see details https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md")]
public interface ISourceGenerator {}
```

**Reply by BowenYang666** [reply]
> It is not ok. Please migrate to IIncrementalGenerator.

OK, I will try this

**Reply by BowenYang666** [reply]
I'm using IIncrementalGenerator now
![image](https://github.com/user-attachments/assets/f29bd4fe-ec3f-42e2-a5ca-c0810d161077)

But when using nuget package reference, the analyzer is gone
![image](https://github.com/user-attachments/assets/ecf21407-21c9-4532-b413-0df8bfe3fbfe)


anything wrong with my  metricFactory.csproj?


<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
	<LangVersion>9.0</LangVersion>
    <GeneratePackageOnBuild>True</GeneratePackageOnBuild>
    <IncludeBuildOutput>false</IncludeBuildOutput> 
  </PropertyGroup>

	<ItemGroup>
		<PackageReference Include="Microsoft.CodeAnalysis.Common" Version="4.11.0" />
		<PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.11.0" PrivateAssets="all" />
		<PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.4" PrivateAssets="all" />
	</ItemGroup>


    <ItemGroup>
    
    <None Include="$(OutputPath)\$(AssemblyName).dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
  </ItemGroup>
</Project>

**Reply by 333fred** [reply]
I don't see anything obviously wrong with that, but that doesn't mean there isn't something wrong. I would do things to troubleshoot next:

1. Look at your actual nuget package, and make sure that the file is ending up where you think it is.
2. Get a binlog of the build for your main project, and verify that the analyzer is actually being passed to the compiler.

What would also be helpful is a repro that I could look at to help.

**Comment by BowenYang666** [comment]
I've put my code here
https://github.com/BowenYang666/MetricFactoryTest/blob/master/NugetTestLocal/MainProject.csproj

and my nuget package is here
[MetricsFactoryTest.1.0.0.zip](https://github.com/user-attachments/files/17123993/MetricsFactoryTest.1.0.0.zip)

**Reply by 333fred** [reply]
Well, once I got your project building (deleting `MetricsFactoryTest\MetricsFactoryTest.sln`, adding a nuget.config to make the built nuget package visible to `NugetLocalTest`), your project built, and did run the generator from the nuget package. So it's something specific to your machine. Did you try getting a binary log and inspecting it?

**Reply by 333fred** [reply]
Also, at a guess, you could have an outdated version in your nuget cache. Try clearing any existing `MetricsFactoryTest` packages from there.

**Reply by BowenYang666** [reply]
I've tried deleted the old MetricsFactoryTest nuget packages. Still cannot work.

This is how I reference the nuget package in my project
![image](https://github.com/user-attachments/assets/14aebd30-4f50-4b6e-8743-c7ec2d99b8a8)
![image](https://github.com/user-attachments/assets/1e05e1f1-9324-4a2e-9b9a-c108b95938ca)

Can you tell me how to add a nuget.config? what does this config looks like, maybe I can try in that way.

And have you sucessfully run the program.cs, mine project will always report a error
![image](https://github.com/user-attachments/assets/ebf7e7c1-f4a6-4a5a-9c10-31bc256429e8)

**Reply by 333fred** [reply]
> I've tried deleted the old MetricsFactoryTest nuget packages

Did you delete it from your nuget cache?
